### ISSUE TITLE
Build fails on SmartOS

### ISSUE THREAD

**Issue by geekrohit** [sass/node-sass#858]
Hi, installing 2.1.1 fails with this output : 

binding.target.mk:128: recipe for target 'Release/obj.target/binding/src/libsass/json.o' failed
make: *** [Release/obj.target/binding/src/libsass/json.o] Error 1
make: Leaving directory '/root/build/bundle/programs/server/npm/lookback:emails/node_modules/node-sass/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/root/build/bundle/programs/server/npm/lookback:emails/node_modules/node-sass/node_modules/pangyp/lib/build.js:271:23)
gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:820:12)
gyp ERR! System SunOS 5.11
gyp ERR! command "/opt/local/bin/node" "/root/build/bundle/programs/server/npm/lookback:emails/node_modules/node-sass/node_modules/pangyp/bin/node-gyp" "rebuild"
gyp ERR! cwd /root/build/bundle/programs/server/npm/lookback:emails/node_modules/node-sass
gyp ERR! node -v v0.10.36
gyp ERR! pangyp -v v2.1.0
gyp ERR! not ok 

gcc version is 4.7.3

Any help would be appreciated!

**Comment by geekrohit** [comment]
And v2.0.0 fails as : 
npm install node-sass@2.0.0
-
> node-sass@2.0.0 install /root/build/bundle/programs/server/npm/lookback:emails/node_modules/node-sass
> node scripts/install.js

Can not download file from https://raw.githubusercontent.com/sass/node-sass-binaries/v2.0.0/sunos-ia32-node-0.10/binding.node

> node-sass@2.0.0 postinstall /root/build/bundle/programs/server/npm/lookback:emails/node_modules/node-sass
> node scripts/build.js


module.js:340
    throw err;
          ^
Error: Cannot find module '/root/build/bundle/programs/server/npm/lookback:emails/node_modules/node-sass/node_modules/pangyp/bin/node-gyp'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:929:3

**Comment by geekrohit** [comment]
Getting same error with 2.0.0 on latest Ubuntu 64bit as well.

**Comment by saper** [comment]
@geekrohit: Can you post the full failure log in the first SmartOS case to gist.github.com?

Regarding the second problem with 2.0.0, can you try to `npm install pangyp` first before installing/building node-sass, I think there was a problem with `package.json`

**Comment by saper** [comment]
Ah I noticed you tried 2.1.1 - can you try [https://github.com/sass/node-sass/tree/v3.0.0-beta.5](3.0.0-beta.5) instead?

**Comment by xzyfer** [comment]
The 2.x releases currently don't falling back to building binaries, and we don't currently build binaries for those versions.

The 3.x version will fallback to building a binary if prebuilt one doesn't exist. Even so there is currently an issue with Libsass regarding SunOS - https://github.com/sass/libsass/issues/850

**Comment by geekrohit** [comment]
For the full failure log : https://gist.github.com/geekrohit/b486551e4483aba6a03d

For 2.0.0, it gives same error of pangyp, even after installing it.

**Comment by geekrohit** [comment]
Applying the patch mentioned in that libsass issue solved the compilation error. Libsass and node-sass works now

**Comment by xzyfer** [comment]
Thanks for giving it a shot @geekrohit. We'll aim to get that patch in before stable

**Comment by geekrohit** [comment]
Ok, great. So currently I had to do two procedures : 
1) Install pangyp inside node-sass module
2) Change uchar_t to uchar32_t in the json.cpp file, libsass library.

After that rebuilding using node scripts/build.js in node-sass directory works fine.

**Comment by xzyfer** [comment]
Great info @geekrohit. These issues will be addressed for 3.0.0-stable

**Comment by djensen47** [comment]
I've tried 3.0.0-pre and I'm getting compilation errors on SmartOS, when will this be addressed? It's seems like you're getting close to stable.

```
$ npm install -g https://github.com/sass/node-sass/tarball/v3.0.0-pre
/home/admin/.nvm/v0.10.38/bin/node-sass -> /home/admin/.nvm/v0.10.38/lib/node_modules/node-sass/bin/node-sass

> node-sass@3.0.0-beta.5 install /home/admin/.nvm/v0.10.38/lib/node_modules/node-sass
> node scripts/install.js

Can not download file from:,https://github.com/sass/node-sass/releases/download/v3.0.0-beta.5/sunos-ia32-11_binding.node

> node-sass@3.0.0-beta.5 postinstall /home/admin/.nvm/v0.10.38/lib/node_modules/node-sass
> node scripts/build.js

Building: /home/admin/.nvm/v0.10.38/bin/node node_modules/pangyp/bin/node-gyp rebuild --libsass_ext= --libsass_cflags= --libsass_ldflags= --libsass_library=
make: Entering directory '/home/admin/.nvm/v0.10.38/lib/node_modules/node-sass/build'
make: *** No rule to make target 'Release/obj.target/libsass/src/libsass/ast.o', needed by 'Release/obj.target/src/sass.a'.  Stop.
make: Leaving directory '/home/admin/.nvm/v0.10.38/lib/node_modules/node-sass/build'
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/admin/.nvm/v0.10.38/lib/node_modules/node-sass/node_modules/pangyp/lib/build.js:271:23)
gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:820:12)
gyp ERR! System SunOS 5.11
gyp ERR! command "/home/admin/.nvm/v0.10.38/bin/node" "/home/admin/.nvm/v0.10.38/lib/node_modules/node-sass/node_modules/pangyp/bin/node-gyp" "rebuild" "--libsass_ext=" "--libsass_cflags=" "--libsass_ldflags=" "--libsass_library="
gyp ERR! cwd /home/admin/.nvm/v0.10.38/lib/node_modules/node-sass
gyp ERR! node -v v0.10.38
gyp ERR! pangyp -v v2.1.0
gyp ERR! not ok
Build failed
```

**Comment by xzyfer** [comment]
@djanowski this fix hasn't been merged yet. Also you want node-sass@3.0.0-beta in until we're stable.

**Comment by djensen47** [comment]
Is there any possibility of getting a patch release fix in for 2.x? This is breaking functionality in other packages that depend on node-sass.

**Comment by saper** [comment]
The file modified, `json.cpp`, is the same in `libsass` 3.1.0 used in node-sass 2.x and in the current `libsass` 3.2.0 series. So just apply https://github.com/saper/libsass/commit/fb43360f1de24044306ccb910a0cddae6f61e136 to your libsass code in the `src/libsass` directory or copy over the resulting [json.cpp](https://raw.githubusercontent.com/saper/libsass/fb43360f1de24044306ccb910a0cddae6f61e136/json.cpp) file.

**Comment by djensen47** [comment]
The issue I'm having is with harp, which depends on terraform, which depends on node-sass, which depends on libsass. I can shrinkwrap but it's going to be a huge pain.

I would think a critical bug would warrant a patch release, no?

**Comment by saper** [comment]
I am not doing releases, so I don't know.

First, it would require to have a libsass 3.1.1 released - the actual node-sass code does not change at all.

What you can do is to patch once, compile node-sass binding and install globally. You should have `binding.node` file in a platform-dependent subdirectory in `vendor`; this can be copied over to other machines etc. If this file is stored under `vendor` `npm install` will simply skip download or compilation steps and will use that file.

In node-sass 3.x you can set an enviroment variable to point to the binary file, but not on 2.x.

**Comment by xzyfer** [comment]
I understand everyone's frustration, and I apologies. Unfortunately at the moment both Libsass and node-sass are managed in a way that would make back ported fixes possible without some major kludging.

We'll have a new beta out in a couple days what will address this issue. Post 3.0.0 stable node-sass will maintain a separate stable branch to make back porting patch easier.

@djanowski I've opened an issue with terraform to bump their dep to node-sass@beta https://github.com/sintaxi/terraform/issues/88

**Comment by geekrohit** [comment]
Also, the issue about pangyp, it exists on both Linux and SunOS platforms.

**Comment by xzyfer** [comment]
If I'm not mistake the pangyp issue was addressed in 2.1.1 and the current beta.

**Comment by geekrohit** [comment]
Ok, then its understood.

**Comment by xzyfer** [comment]
This has been resolved in 3.0.0-beta.7.

```
npm install node-sass@beta
```
